pkg/blockbuilder/scheduler/priority_queue.go

```diff
@@ -4,91 +4,158 @@ import (
 	"container/heap"
 )
 
-// PriorityQueue is a generic priority queue.
-type PriorityQueue[T any] struct {
-	h *priorityHeap[T]
+// PriorityQueue is a generic priority queue with constant time lookups.
+type PriorityQueue[K comparable, V any] struct {
+	h   *priorityHeap[V]
+	m   map[K]*item[V] // Map for constant time lookups
+	key func(V) K      // Function to extract key from value
+}
+
+// item represents an item in the priority queue with its index
+type item[V any] struct {
+	value V
+	index int
 }
 
 // NewPriorityQueue creates a new priority queue.
-func NewPriorityQueue[T any](less func(T, T) bool) *PriorityQueue[T] {
-	h := &priorityHeap[T]{
+func NewPriorityQueue[K comparable, V any](less func(V, V) bool, key func(V) K) *PriorityQueue[K, V] {
+	h := &priorityHeap[V]{
 		less: less,
-		heap: make([]T, 0),
+		heap: make([]*item[V], 0),
+		idx:  make(map[int]*item[V]),
 	}
 	heap.Init(h)
-	return &PriorityQueue[T]{h: h}
+	return &PriorityQueue[K, V]{
+		h:   h,
+		m:   make(map[K]*item[V]),
+		key: key,
+	}
 }
 
 // Push adds an element to the queue.
-func (pq *PriorityQueue[T]) Push(v T) {
-	heap.Push(pq.h, v)
+func (pq *PriorityQueue[K, V]) Push(v V) {
+	k := pq.key(v)
+	if existing, ok := pq.m[k]; ok {
+		// Update existing item's value and fix heap
+		existing.value = v
+		heap.Fix(pq.h, existing.index)
+		return
+	}
+
+	// Add new item
+	idx := pq.h.Len()
+	it := &item[V]{value: v, index: idx}
```

Comment:
there is a bit of duplication, we update the `index` of the item here and within the heap's [Push()](https://github.com/grafana/loki/pull/15338/files#diff-fdfb40c0504795e56a339b2c872bdb609e2d29a6a493968a70ee447a5db66a11R124) method.